dex/src/main/scala/com/wavesplatform/dex/api/MatcherApiRoute.scala

```diff
@@ -194,19 +196,34 @@ case class MatcherApiRoute(assetPairBuilder: AssetPairBuilder,
                            required = true)
     )
   )
-  def upsertRate: Route = (path("settings" / "rates" / AssetPM) & put & withAuth) { a =>
-    entity(as[Double]) { rate =>
-      withAsset(a) { asset =>
-        complete(
-          if (asset == Waves) StatusCodes.BadRequest -> wrapMessage("Rate for Waves cannot be changed")
-          else {
-            val assetStr = AssetPair.assetIdStr(asset)
-            rateCache.upsertRate(asset, rate) match {
-              case None     => StatusCodes.Created -> wrapMessage(s"Rate $rate for the asset $assetStr added")
-              case Some(pv) => StatusCodes.OK      -> wrapMessage(s"Rate for the asset $assetStr updated, old value = $pv, new value = $rate")
+  def upsertRate: Route = {
+
+    val rateParsingRejectionsHandler =
```

Comment:
Could you move this outside the `def` to avoid initialization on each call?

Reply:
Done